_drafts/test-automation-renode.md

```diff
@@ -0,0 +1,886 @@
+```
+
+### Complete Github Action `main.yml`
+
+```yaml
+name: Renode Automated Tests
+
+on:
+  push:
+    branches: [master]
+  pull_request:
+    branches: [master]
+
+jobs:
+  build_and_test:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          submodules: true
+
+      # Get the arm-non-eabi-gcc toolchain
+      - name: Install arm-none-eabi-gcc
+        uses: fiam/arm-none-eabi-gcc@v1
+        with:
+          # The arm-none-eabi-gcc release to use.
+          release: "9-2019-q4"
+
+      - name: Build Firmware
+        run: make -j4
+
+      - name: Upload ELF
+        uses: actions/upload-artifact@v2
+        with:
+          name: renode-example.elf
+          path: build/renode-example.elf
+
+      - name: Renode Tests
+        run: ./docker-test.sh
+
+      - name: Upload Output Dir
+        uses: actions/upload-artifact@v2
+        with:
+          name: Renode Test Results
+          path: test_results/
+```
+
+Now, when we publish a pull-request on Github, we'll immediately see that the
+build is triggered and our job starts.
+
+![](/img/test-automation-renode/github-pr-building.png)
+
+If we click on "Details", we can watch each job complete in real time! Below we
+see that all jobs are successful and that we have two artifacts. One is the ELF
+file that was built during the job, and the other is a ZIP archive of the Robot
+Framework test results.
+
+![](/img/test-automation-renode/github-pr-build.png)
+
+The best part about using the Robot Framework integration of Renode is that it
+generates pretty HTML-based reports, as we saw in the [Introduction to Renode
+post]({% post_url 2020-03-23-intro-to-renode %}#renode--integration-tests).
+
+> I've included the HTML report for the above test.
+> [Click here to view it](/misc/test-automation-renode/log.html).
+
+## Debugging Failing Tests in CI
+
+Tests are going to fail in CI, and it's probably a good thing, as that's what CI
+is for. But, it would be a pain to guess and check how to fix issues that only
+occur in CI. Thankfully, Renode has the ability to capture the state of the
+system, save it to a file, and load it after-the-fact[^renode_state_save].
+
+It does this using the monitor commands `Save` and `Load`. Wouldn't it be cool
+if we could call the `Save` command in CI for failing tests and then locally run
+`Load` in Renode to get the exact state of the device at the time of failure?
+Yes it would be.
+
+Renode has part of this
+[built-in](https://github.com/renode/renode/blob/05377ef375daa3d5ea0de12633d27bf26e20b3b3/src/Renode/RobotFrameworkEngine/renode-keywords.robot#L77-L86).
+Unfortunately, this is yet another piece of functionality that doesn't exist in
+the current public release, so I've
+[copied the code](https://github.com/memfault/interrupt-renode-test-automation/blob/master/tests/common.robot)
+(and overrode the snapshot save location) so that these snapshots get uploaded
+into our Github Actions artifacts!
```

Comment:
Did you have to configure Github actions to do this? Is there a special path that is used for artifacts?